spec/models/miq_queue_spec.rb

```diff
@@ -6,13 +6,132 @@
       _, @miq_server, @zone = EvmSpecHelper.create_guid_miq_server_zone
     end
 
-    it "works with deliver_on" do
+    it "requires class_name" do
+      msg = MiqQueue.new(:class_name => nil)
+      status, message, result = msg.deliver
+      expect(status).to eq(MiqQueue::STATUS_ERROR)
+      expect(message).to eq("class_name cannot be nil")
+      expect(result).to be_nil
+    end
+
+    it "requires valid class_name" do
+      msg = MiqQueue.new(:class_name => "NotARealClazz")
+      status, message, result = msg.deliver
+      expect(status).to eq(MiqQueue::STATUS_ERROR)
+      expect(message).to eq("uninitialized constant NotARealClazz")
+      expect(result).to be_nil
+    end
+
+    it "uses class_name without instance_id" do
+      expect(MiqServer).to receive(:my_zone).and_return("MY ZONE")
+      msg = MiqQueue.new(:class_name => "MiqServer", :method_name => "my_zone")
+
+      status, message, result = msg.deliver
+      expect(status).to eq(MiqQueue::STATUS_OK)
+      expect(message).to eq("Message delivered successfully")
+      expect(result).to eq("MY ZONE")
+    end
+
+    it "uses object with instance_id" do
+      expect(MiqServer).to receive(:find).with(@miq_server.id).and_return(@miq_server)
+      expect(@miq_server).to receive(:my_zone).and_return("MY ZONE")
+      msg = MiqQueue.new(:class_name => "MiqServer", :instance_id => @miq_server.id, :method_name => "my_zone")
+
+      status, message, result = msg.deliver
+      expect(status).to eq(MiqQueue::STATUS_OK)
+      expect(message).to eq("Message delivered successfully")
+      expect(result).to eq("MY ZONE")
+    end
+
+    it "handles record not found" do
+      msg = MiqQueue.new(:class_name => "MiqServer", :instance_id => @miq_server.id + 100, :method_name => "my_zone")
+
+      expect(msg._log).to receive(:warn).with(/will not be delivered because/)
+
+      status, message, result = msg.deliver
+      expect(status).to eq(MiqQueue::STATUS_WARN)
+      expect(message).to be_nil
+      expect(result).to be_nil
+    end
+
+    it "handles random AR error" do
+      msg = MiqQueue.new(:class_name => "MiqServer", :instance_id => @miq_server.id, :method_name => "my_zone")
+
+      expect(MiqServer).to receive(:find).with(@miq_server.id).and_raise("ugh")
```

Comment:
Is this the same test as the timeout error one below but generically for any error?  Can cheat and do

```ruby
allow(MiqServer).to receive(:bomb).and_raise(StandardError, "ugh!!!")
msg = MiqQueue.new(:class_name => "MiqServer", :method_name => "bomb")

```

Reply:
@jrafanie no, this is targeting `MiqServer.find()` when `:instance_id` is specified. [[here]](https://github.com/ManageIQ/manageiq/blob/6e5fb5083a8a03299ce2cef72e7cfb8f3a179570/app/models/miq_queue.rb#L407,L411)

The other :bomb: is not `instance_id` based, but rather the actual method based [[here]](https://github.com/ManageIQ/manageiq/blob/6e5fb5083a8a03299ce2cef72e7cfb8f3a179570/app/models/miq_queue.rb#L425-L434) (sorry - ignore the `MiqQueueRetryLater` bit)

The goal of fully spec'ing this out is so I can simplify that method.
So if you think I can simplify these, that would give us more flexibility in refactoring that method

Reply:
I guess I don't want to mock find 😦

Reply:
@jrafanie Yea. mocking `find` is not pretty. Forcing a `RecordNotFound` (block `#A`) from `find` is easy. Not sure how to force another error (block `#B`). I'll try and find another way, but this feels like splitting hairs a little.

```ruby
begin
  obj = obj.find(instance_id)
rescue ActiveRecord::RecordNotFound => err
  # A easy
rescue => err
  # B goal of this test
end
```

Reply:
I don't know, maybe don't test this?